chips/stm32f303xc/src/adc.rs

```diff
@@ -0,0 +1,740 @@
+    Channel2 = 0b00010,
+    Channel3 = 0b00011,
+    Channel4 = 0b00100,
+    Channel5 = 0b00101,
+    Channel6 = 0b00110,
+    Channel7 = 0b00111,
+    Channel8 = 0b01000,
+    Channel9 = 0b01001,
+    Channel10 = 0b01010,
+    Channel11 = 0b01011,
+    Channel12 = 0b01100,
+    Channel13 = 0b01101,
+    Channel14 = 0b01110,
+    Channel15 = 0b01111,
+    Channel16 = 0b10000,
+    Channel17 = 0b10001,
+    Channel18 = 0b10010,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum DiscontinuousMode {
+    OneChannels = 0b000,
+    TwoChannels = 0b001,
+    ThreeChannels = 0b010,
+    FourChannels = 0b011,
+    FiveChannels = 0b100,
+    SixChannels = 0b101,
+    SevenChannels = 0b110,
+    EightChannels = 0b111,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum ExternalTriggerDetection {
+    Disabled = 0b00,
+    RisingEdge = 0b01,
+    FallingEdge = 0b10,
+    RisingAndFalling = 0b11,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum ExternalTriggerSelection {
+    Event0 = 0b0000,
+    Event1 = 0b0001,
+    Event2 = 0b0010,
+    Event3 = 0b0011,
+    Event4 = 0b0100,
+    Event5 = 0b0101,
+    Event6 = 0b0110,
+    Event7 = 0b0111,
+    Event8 = 0b1000,
+    Event9 = 0b1001,
+    Event10 = 0b1010,
+    Event11 = 0b1011,
+    Event12 = 0b1100,
+    Event13 = 0b1101,
+    Event14 = 0b1110,
+    Event15 = 0b1111,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum DataResolution {
+    Bit12 = 0b00,
+    Bit10 = 0b01,
+    Bit8 = 0b10,
+    Bit6 = 0b11,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum ADCStatus {
+    Idle,
+    Off,
+    PoweringOn,
+    OneSample,
+}
+
+pub struct Adc {
+    registers: StaticRef<AdcRegisters>,
+    common_registers: StaticRef<AdcCommonRegisters>,
+    clock: AdcClock,
+    status: Cell<ADCStatus>,
+    client: OptionalCell<&'static dyn EverythingClient>,
+    // clock34: AdcClock,
+}
+
+pub static mut ADC1: Adc = Adc::new();
+
+impl Adc {
+    const fn new() -> Adc {
+        Adc {
+            registers: ADC1_BASE,
+            //rcc copy
+            common_registers: ADC12_COMMON_BASE,
+            clock: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC1)),
+            // clock34: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC34)),
+            status: Cell::new(ADCStatus::Off),
+            client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn enable(&self) {
+        self.status.set(ADCStatus::PoweringOn);
+
+        // Enable adc clock
+        self.enable_clock();
+
+        //Set Synchronous clock mode
+        self.common_registers.ccr.modify(CCR::CKMODE.val(0b01));
+
+        self.registers.cr.modify(CR::ADVREGEN.val(0b00));
+        self.registers.cr.modify(CR::ADVREGEN.val(0b01));
+
+        // Wait for ADVRGEN to enable
+        // This needs to be synchronous because there is no interrupt signaling
+        // when ADVRGEN becomes enabled
+        for _i in 0..1000000 {
+            unsafe {
+                llvm_asm!(
+                "nop"
+            : : : : "volatile" );
+            }
+        }
```

Comment:
is there a bit we can check that will indicate when ADVRGEN has been enabled? Looping for a million cycles seems a shame if we can short circuit once it becomes ready, and it would be nice to not add `llvm_asm` to this crate if we don't have to.

Reply:
No, it just needs a wait time for it to be enabled and to continue the ADC powering on

Reply:
This is the link to the reference manual for stm32f3xx https://www.st.com/resource/en/reference_manual/dm00043574-stm32f303xb-c-d-e-stm32f303x6-8-stm32f328x8-stm32f358xc-stm32f398xe-advanced-arm-based-mcus-stmicroelectronics.pdf and on page 316 it is described how to enable ADVREGEN. Am I missing anything? I’m open to any ideas.

Reply:
That seems to indicate the worst case delay is 10 microseconds. Also, my understanding is that the maximum core clock speed is 72 MHz on this chip, so I think the worst case number of cycles should be 720. Can you replace 1000000 with 720, and add a comment explaining why that number is used? 

Ideally we would have a safe wait primitive like that described in #1552 , but for now I think limiting this to 720 cycles should reflect the spirit of that.